Automattic/sensei
# Add setting in user profile page to Opt out of or Opt in to receiving individual emails

Resolves https://github.com/Automattic/sensei/issues/7497

## Proposed Changes

Added a setting section in the User profile page in WpAdmin where the user can Opt out of (Or later Opt in to) individual emails they receive.

These were the considerations - 

- Any student/teacher/user can go to their own profile and uncheck any box beside an email to Opt-out
- If any new email is introduced, by default, the users will subscribe to them, because our emails are Opt-out by nature
- If the Admin keeps any email deactivated, the users/students will not see in in their profile, because they don’t need to opt out
- Students will see only the Student emails in their profile page, because seeing Teacher emails don’t make sense for them
- Pro emails will only be available in users’ Profile if Pro is installed, we don’t show the upsell emails there, because it doesn’t make sense for the users to opt-out of unavailable emails
- An admin user (who has Cap for editing profiles of other users) can open a user’s profile from WpAdmin -> Users -> All Users -> (Click on any user) and opt them out/in to any email. In case that site has WP Admin disabled, or any user has asked the teacher/admin to stop getting a particular email,  the Admin can just do it for that particular user (Can be handy in cases where users don’t have access to WpAdmin)

Note: We initially planned to add a link in the bottom of the emails with an URL to this page so that the users will know where to go to for unsubscribing. But @donnapep and I discussed later and decided not to do it here. Because in some sites, users are not given access to WP-Admin, putting this URL in the emails of those sites will not make sense.

## Testing Instructions


1. Make sure you are receiving Sensei emails for your Admin user
2. Create a User with a student role, with another email you can check
3. Create a Course with lessons and quizzes
4. Enrol the student User to that Course
5. Do some email generating activities (Enrolling in course should do, can also submit quiz by logging in as the student and the teacher user can grade the quiz)
6. Make sure the student receives the emails
7. Now after logging in as the student user, go to Wp Admin -> Profile
8. Scroll down to **Sensei Email** section
9. Make sure you are seeing emails that are only for Student users, no Teacher Email is appearing
10. Opt out of one or more emails that you were receiving (uncheck and save)
11. Now generate those emails again
12. Make sure you haven't received any
13. Now turn those emails on again
14. Regenerate emails
15. Make sure you are receiving them now
16. Log in as the Admin user and go to Users -> Profile (your own Admin user's profile)
17. Make sure you are seeing all active emails
18. Opt out and in and check by email receiving as before
19. As an admin user, go to Users -> WpAdmin -> Users -> All Users -> (Click on the student user) and opt the user out of any email
20. Generate email for student user and make sure the logics work
21. Deactivate Senesi-Pro and make sure Pro emails don't appear in profiles (Or activate and make sure they do)
22. Unpublish one or more emails and make sure they don't appear in profiles any more
23. Re-activate and make sure they are appearing again
24. Login with users of different roles and make sure you see this setting in their profiles

## New/Updated Hooks


* `sensei_send_emails` - Updated hook by adding a new param to identify the email being sent.

## Pre-Merge Checklist
